File: packages/client-python/src/sbp/journal.py
"""
Append-only JSONL operation journal (P3.1): one line per blackboard operation.
"""
import json
import os
from typing import Any


class Journal:
    def __init__(self, path: str | os.PathLike[str]) -> None:
        self._path = path
        self._seq = 0
        dirname = os.path.dirname(path)
        if dirname:
            os.makedirs(dirname, exist_ok=True)

    def append(self, entry: dict[str, Any]) -> None:
        entry = {"seq": self._seq, **entry}
        self._seq += 1
        try:
            with open(self._path, "a") as f:
                f.write(json.dumps(entry, separators=(",", ":")) + "\n")
        except OSError as e:
            print(f"[SBP Journal] write failed: {e}")

File: packages/server/src/journal.ts
import { appendFileSync, mkdirSync } from "fs";
import { dirname } from "path";

// Append-only JSONL op journal; seq is 1-based and gapless per file.
// A failed append is logged, never raised — journaling must not break the op it records.
export class Journal {
  private seq = 0;

  constructor(private readonly path: string) {
    const dir = dirname(path);
    if (dir && dir !== ".") mkdirSync(dir, { recursive: true });
  }

  append(entry: Record<string, unknown>): void {
    const line = { seq: ++this.seq, ...entry };
    try {
      appendFileSync(this.path, JSON.stringify(line) + "\n");
    } catch (e) {
      console.error(`[SBP Journal] write failed: ${e}`);
    }
  }
}
